File: human-solutions/06/python/main.py
from pathlib import Path

input_text = Path("../../../inputs/06.txt").read_text().rstrip()


def solve_part1(input_text: str) -> int:
    rows = [
        line.strip().split()
        for line in input_text.splitlines()
    ]

    number_rows = [
        list(map(int, row))
        for row in rows[:-1]
    ]
    operators = rows[-1]

    columns = [
        [row[i] for row in number_rows]
        for i in range(len(operators))
    ]

    def apply_op(values, op):
        result = values[0]
        for v in values[1:]:
            if op == "+":
                result += v
            elif op == "*":
                result *= v
            else:
                raise ValueError(f"Unknown operator: {op}")
        return result

    column_results = [
        apply_op(col, operators[i])
        for i, col in enumerate(columns)
    ]

    return sum(column_results)


def solve_part2(input_text: str) -> int:
    lines = input_text.splitlines()
    operator_row = lines.pop()
    digit_rows = lines

    height = len(digit_rows)
    width = max(
        len(operator_row),
        *(len(r) for r in digit_rows)
    )

    total = 0
    current_numbers = []
    current_op = None

    def flush_problem():
        nonlocal total, current_numbers, current_op
        if not current_numbers or current_op is None:
            return

        if current_op == "+":
            value = sum(current_numbers)
        else:
            value = 1
            for n in current_numbers:
                value *= n

        total += value
        current_numbers = []
        current_op = None

    for col in range(width - 1, -1, -1):
        digits = []

        for row in range(height):
            if col < len(digit_rows[row]):
                ch = digit_rows[row][col]
                if ch != " ":
                    digits.append(ch)

        op = operator_row[col] if col < len(operator_row) else " "

        is_blank = not digits and op == " "

        if is_blank:
            flush_problem()
            continue

        if op in "+*":
            current_op = op

        if digits:
            current_numbers.append(int("".join(digits)))

    flush_problem()
    return total


print(solve_part1(input_text))
print(solve_part2(input_text))

File: human-solutions/06/rust/src/main.rs
use std::fs;

fn solve_part1(input: &str) -> i64 {
    let rows: Vec<Vec<&str>> = input
        .lines()
        .map(|l| l.trim().split_whitespace().collect())
        .collect();

    let number_rows: Vec<Vec<i64>> = rows[..rows.len() - 1]
        .iter()
        .map(|row| row.iter().map(|v| v.parse().unwrap()).collect())
        .collect();

    let operators = &rows[rows.len() - 1];

    let mut total = 0;

    for col in 0..operators.len() {
        let mut values: Vec<i64> = Vec::new();
        for row in &number_rows {
            values.push(row[col]);
        }

        let mut result = values[0];
        for v in values.iter().skip(1) {
            match operators[col] {
                "+" => result += v,
                "*" => result *= v,
                _ => panic!("Unknown operator"),
            }
        }

        total += result;
    }

    total
}

fn solve_part2(input: &str) -> i64 {
    let mut lines: Vec<&str> = input.lines().collect();
    let operator_row = lines.pop().unwrap();
    let digit_rows = lines;

    let height = digit_rows.len();
    let width = std::iter::once(operator_row.len())
        .chain(digit_rows.iter().map(|r| r.len()))
        .max()
        .unwrap();

    let mut total: i64 = 0;
    let mut current_numbers: Vec<i64> = Vec::new();
    let mut current_op: Option<char> = None;

    let flush_problem = |total: &mut i64,
                             nums: &mut Vec<i64>,
                             op: &mut Option<char>| {
        if nums.is_empty() || op.is_none() {
            return;
        }

        let value: i64 = match op.unwrap() {
          '+' => nums.iter().sum::<i64>(),
          '*' => nums.iter().product::<i64>(),
          _ => unreachable!(),
        };

        *total += value;
        nums.clear();
        *op = None;
    };

    for col in (0..width).rev() {
        let mut digits: Vec<char> = Vec::new();

        for row in 0..height {
            if col < digit_rows[row].len() {
                let ch = digit_rows[row].as_bytes()[col] as char;
                if ch != ' ' {
                    digits.push(ch);
                }
            }
        }

        let op = if col < operator_row.len() {
            operator_row.as_bytes()[col] as char
        } else {
            ' '
        };

        let is_blank = digits.is_empty() && op == ' ';

        if is_blank {
            flush_problem(&mut total, &mut current_numbers, &mut current_op);
            continue;
        }

        if op == '+' || op == '*' {
            current_op = Some(op);
        }

        if !digits.is_empty() {
            let number: i64 = digits.iter().collect::<String>().parse().unwrap();
            current_numbers.push(number);
        }
    }

    flush_problem(&mut total, &mut current_numbers, &mut current_op);
    total
}

fn main() {
    let input = fs::read_to_string("../../../inputs/06.txt").unwrap();
    let input = input.trim_end();

    println!("{}", solve_part1(input));
    println!("{}", solve_part2(input));
}
